src/settings/index.tsx

```diff
@@ -1,43 +1,112 @@
+import { ipcRenderer } from "electron";
+import { settingsType, Bookmark } from "../settings";
+import { CaptureMode } from "../native";
 import * as React from "react";
 import * as ReactDom from "react-dom";
 
-
-
 import "./style.scss";
 import "./index.html";
 
+window.addEventListener("DOMContentLoaded", start);
 
-type Alt1Settings = ReturnType<typeof defaultSettings>
+async function start() {
+	let settings: settingsType = await ipcRenderer.invoke("getsettings");
+	ReactDom.render(<Settings settings={settings} />, document.getElementById("root"));
+}
 
-function defaultSettings() {
-	return {
-		captMode: "opengl" as "opengl" | "directx" | "desktop",
-		captAutoDetect: true
-	};
+interface SettingsProps {
+	settings: settingsType;
 }
 
+interface SettingsState {
+	tab: string;
+}
 
-window.addEventListener("DOMContentLoaded", start);
+class Settings extends React.Component<SettingsProps, SettingsState> {
+	constructor(props) {
+		super(props);
+		this.state = {
+			tab: "apps"
+		}
+	}
+
+	render() {
+		let content = <h1>Invalid tab</h1>;
+
+		if (this.state.tab == "apps") {
+			content = <AppSettings bookmarks={this.props.settings.bookmarks}/>
+		} else if (this.state.tab == "capture") {
+			content = <CaptureSettings captureMode={this.props.settings.captureMode}/>
+		}
+
+		return (
+			<React.Fragment>
+				<button onClick={() => this.setState({tab: "apps"})}>Apps</button>
+				<button onClick={() => this.setState({tab: "capture"})}>Capture</button>
+				<hr/>
+				{content}
+			</React.Fragment>
+		);
+	}
+}
 
-let settings = defaultSettings();
+interface AppSettingsProps {
+	bookmarks: Array<Bookmark>;
+}
 
-function start() {
-	ReactDom.render(<Settings />, document.getElementById("root"));
+interface AppSettingsState {
+	configUrl: string;
 }
 
+class AppSettings extends React.Component<AppSettingsProps, AppSettingsState> {
+	constructor(props) {
+		super(props);
+		this.state = {
+			configUrl: ""
+		}
+	}
+
+	addAppSubmit(e) {
+		ipcRenderer.invoke("installapp", this.state.configUrl).then(() => start());
+		e.preventDefault();
```

Comment:
Ideally, the redraw trigger should be an event trigger in the main process. When that is implemented correctly you only have to send the commands and no longer worry about ui state.